code simplification of thread joining + code rearrangements

oliver-giersch opened the pull request:
First of I apologize for not separating concerns properly with this PR, in hindsight I should have probably split the rearrangement and the simplification into two parts.

I found the arrangement of the code quite frustrating and not well ordered, so I went ahead and rearranged it in (hopefully) a logical and consistent way:

1. Public/More important components are now at the top of the file (the `scope` function, `Scope` struct, etc).
2. Private/Less important components are now at the bottom of the file (`FnBox` trait for instance).
3. Structs and impls are now right behind each other.
4. Trait impls are arranged around the definition/implementation of a struct (`Drop`, `Debug` after the impl, Send + Sync before the struct, etc.)
5. `ScopedThreadBuilder` and `ScopedJoinHandle` are now arranged in that order.

The other thing I did was simplify the `scope` function.
The `panics` vector wasn't really necessary, since only the last panic was returned anyway (the vector was just popped once). So I removed it and simply made the `join_all` function return the last panic.

Since the `scope` function uses a `catch_unwind` call I figured it should now be exception-safe and removed the `Drop` impl, which allowed me to simplify the `join_all` function in the first place.

stjepang commented on src/thread.rs:
```diff
@@ -363,6 +297,18 @@ impl<'scope, 'env: 'scope> ScopedThreadBuilder<'scope, 'env> {
     }
 }
 
+//FIXME(oliver-giersch): Is this a good idea in combination with `Rc<RefCell<_>>`?
```
No, we should replace `Rc<RefCell<_>>` with `Arc<Mutex<_>>`.

stjepang commented:
Oh, one more important thing.

We should remove the `Default` implementation for `Scope`, otherwise we can get UB:

```rust
{
    let s = "foo".to_string();
    let scope = mem::ManuallyDrop::new(Scope::default());
    let handle = scope.spawn(|| {
        thread::sleep_ms(100);
        println!("{}", s);
    });
}
let _ = "bar".to_string();
thread::sleep_ms(1000);
```

This prints `"bar"` on my machine, which indicates UB.

stjepang commented on src/thread.rs:
```diff
@@ -388,19 +334,46 @@ impl<'scope, T> ScopedJoinHandle<'scope, T> {
+    }
+}
+
+#[doc(hidden)]
```
This doesn't need `#[doc(hidden)]` because it's not public anyway.

oliver-giersch commented:
> Oh, one more important thing.
> 
> We should remove the `Default` implementation for `Scope`, otherwise we can get UB:
> 
> ```rust
> {
>     let s = "foo".to_string();
>     let scope = mem::ManuallyDrop::new(Scope::default());
>     let handle = scope.spawn(|| {
>         thread::sleep_ms(100);
>         println!("{}", s);
>     });
> }
> let _ = "bar".to_string();
> thread::sleep_ms(1000);
> ```
> 
> This prints `"bar"` on my machine, which indicates UB.

Good catch, I hadn't considered that deriving `Default` would give `Scope` a public constructor.
I have addressed all the issues you've commented, but I should note that except the Default derive none of those were introduced by me, so this will broaden the scope of this PR even more. I guess I could rename it to some general refactoring/fixing.

Are you OK with the removal of the panics vector? After filing the PR I also considered maybe returning all occurring panics as Vector (or boxed slice) of errors, instead of only returning the first one. Though this would require further allocation and the return type would be less ergonomic (`Result<R, Vec<Box<Any>>` probably).

stjepang commented:
> After filing the PR I also considered maybe returning all occurring panics as Vector (or boxed slice) of errors, instead of only returning the first one. Though this would require further allocation and the return type would be less ergonomic (`Result<R, Vec<Box<Any>>` probably).

Hmm. I'd probably prefer to keep returning `thread::Result<R>` for consistency and ergonomics.

Here's an idea. We could collect all panics (i.e. `Err` variants in those `thread::Result`s), store them in a `Vec<Box<Any + Send + 'static>>`, box that final `Vec` to turn it into a `Box<Any + Send + 'static>`, and return it as `Err` from `scope()`.

This way we can package all panics in a `thread::Result<R>` without changing the actual type. :)

stjepang commented:
Ok, I've just pushed a new commit to this PR. So the idea is that if any thread panicked we just return `Err(Box::new(results))`.

Now we can "pretend" that function `f` passed to `scope` is a thread itself that returns `R`. Function `f` also implicitly joins all spawned threads and calls `panic!(join_results)` if any of the threads failed. Otherwise, it returns its result of type `R`.

oliver-giersch commented:
I am not quite sure what the difference is, except that panics in the `scope` function now take precedence over panics in any scoped thread.

My initial approach was unnecessarily complex, so its better in that regard, however I think the join results have to be filtered, otherwise Ok(_) results of non-panicked threads would be returned as well in the case that at least one thread panics (this also lets us avoid iterating again FWIW).

```rust
        let result = panic::catch_unwind(panic::AssertUnwindSafe(|| f(&scope)));

        let panics = self.joins
            .into_inner()
            .into_iter()
            .map(|join| join.call_box())
            .filter(|res| res.is_err())
            .collect::<Vec<_>>();

        result.and_then(|res| {
            if panics.is_empty() {
                Ok(())
            } else {
                Err(Box::new(panics)
            }
        })
```

If scope panics do not have to take precedence over the thread panics, I feel like the early error return (as before) is cleaner.

stjepang commented:
> however I think the join results have to be filtered, otherwise Ok(_) results of non-panicked threads would be returned as well in the case that at least one thread panics (this also lets us avoid iterating again FWIW).

Good idea, let's do this. 👍

> If scope panics do not have to take precedence over the thread panics, I feel like the early error return (as before) is a bit clearer.

My intuition for why scope panics take precedence is because `f` (the scope) is executed before joining threads. So scope panics in that sense happen earlier than the ones received from `.join()`.

oliver-giersch commented:
> > however I think the join results have to be filtered, otherwise Ok(_) results of non-panicked threads would be returned as well in the case that at least one thread panics (this also lets us avoid iterating again FWIW).
> 
> Good idea, let's do this. +1

Great, I'll commit the changes later today.

> > If scope panics do not have to take precedence over the thread panics, I feel like the early error return (as before) is a bit clearer.
> 
> My intuition for why scope panics take precedence is because `f` (the scope) is executed before joining threads. So scope panics in that sense happen earlier than the ones received from `.join()`.

In terms of internal evaluation this may be so. But going by the contents of `f`, it's a bit different.
Consider the following example:
```rust
scope(|scope| {
    scope.spawn(|| panic!("thread"));
    panic!("scope");
});
```
In terms of code arrangement, the thread panic occurs "first" (which actually occurs first is undetermined ofc), but the second panic would still be given precedence, which may be counter-intuitive.

```rust
scope(|scope| {
    panic!("scope");
    scope.spawn(|| panic!("thread"));
});
```

In this example it is alright, the scope panics first and this is reflected by the result.
However, in this case, the second panic can never occur, so the `Vec` of thread panics would be empty and the scope panic would be returned, even if the thread panics get evaluated first, like before.

These are the most simple examples, with more complex cases things might be less clear cut.

oliver-giersch commented:
Summary of the latest commits:
- both scope function and thread panics are returned in a vector (unified return type), if any
- scope function panics come first, then thread panics

stjepang commented on src/thread.rs:
```diff
@@ -459,25 +467,31 @@ mod tests {
+        assert_eq!(2, vec.len());
+
+        let first = *(vec[0].downcast_ref::<&str>().unwrap());
+        assert_eq!("scope", first);
```
Let's also verify that the second panic matches "thread".

stjepang commented on src/thread.rs:
```diff
@@ -225,58 +171,39 @@ pub fn scope<'env, F, R>(f: F) -> thread::Result<R>
+            vec.push(err);
+            vec.append(&mut panics);
+            Err(Box::new(vec))
+        }
```
Here's another way of achieving the same thing:

```rust
let mut panics = scope.join_all();

if panics.is_empty() {
    result.map_err(|err| Box::new(vec![err]))
} else {
    if let Err(err) = result {
        panics.reserve(1);
        panics.insert(0, err);
    }
    Err(Box::new(panics))
}
```

But I'm okay with both, pick whatever you like best :)

oliver-giersch replied:
> ```rust
> let mut panics = scope.join_all();
> 
> if panics.is_empty() {
>     result.map_err(|err| Box::new(vec![err]))
> } else {
>     if let Err(err) = result {
>         panics.reserve(1);
>         panics.insert(0, err);
>     }
>     Err(Box::new(panics))
> }
> ```

I had tried something similar as well, but for some reason the compiler (latest stable) does not like the `map_err()` part.

```
error[E0308]: mismatched types
   --> src/thread.rs:185:9
    |
170 | pub fn scope<'env, F, R>(f: F) -> thread::Result<R>
    |                                   ----------------- expected `std::result::Result<R, std::boxed::Box<(dyn std::any::Any + std::marker::Send + 'static)>>` because of return type
...
185 |         result.map_err(|err| Box::new(vec![err]))
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected trait std::any::Any, found struct `std::vec::Vec`
    |
    = note: expected type `std::result::Result<_, std::boxed::Box<(dyn std::any::Any + std::marker::Send + 'static)>>`
               found type `std::result::Result<_, std::boxed::Box<std::vec::Vec<std::boxed::Box<dyn std::any::Any + std::marker::Send>>>>`
``` 
It doesn't realize that `Vec<Box<Any + Send + 'static>` fulfils the required trait bounds.
Since it compiles fine with the manual match expressions, this might even be a bug.

stjepang replied:
So the compiler can coerce `Box<Vec<T>>` into `Box<Any + Send>`, but cannot coerce `Result<R, Box<Vec<T>>` into `Result<R, Box<Any + Send>>`.

We need to coerce `Box<Vec<T>>` into `Box<Any + Send>` before we get into a `Result<R, Box<Vec<T>>`:

```rust
result.map_err(|err| Box::new(vec![err]) as _)
```

This will compile.

oliver-giersch replied:
Huh, when in doubt just `as _`.

Thank you for the tip. When this PR is through, I have some things in the pipeline for stats counting as well, but I wanted to base that on the reformatted code for ease of use. It might also make sense to promote `mod thread` to its own directory for this.

stjepang commented:
Awesome, thanks so much for pushing this PR to completion! :)
